src/WebJobs.Script.WebHost/Security/Authentication/AuthenticationLevelHandler.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Azure.WebJobs.Extensions.Http;
+using Microsoft.Azure.WebJobs.Script.WebHost.Features;
+using Microsoft.Azure.WebJobs.Script.WebHost.Security.Authentication;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Authentication
+{
+    internal class AuthenticationLevelHandler : AuthenticationHandler<AuthenticationLevelOptions>
+    {
+        public const string FunctionsKeyHeaderName = "x-functions-key";
+        private readonly ISecretManager _secretManager;
+
+        public AuthenticationLevelHandler(
+            IOptionsMonitor<AuthenticationLevelOptions> options,
+            ILoggerFactory logger,
+            UrlEncoder encoder,
+            IDataProtectionProvider dataProtection,
+            ISystemClock clock,
+            ISecretManager secretManager)
+            : base(options, logger, encoder, clock)
+        {
+            _secretManager = secretManager;
+        }
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            // Get the authorization level for the current request
+            (string name, AuthorizationLevel requestAuthorizationLevel) = await GetAuthorizationKeyInfoAsync(Context.Request, _secretManager);
+
+            if (requestAuthorizationLevel != AuthorizationLevel.Anonymous)
+            {
+                var claims = new List<Claim>
+                {
+                    new Claim(SecurityConstants.AuthLevelClaimType, requestAuthorizationLevel.ToString())
+                };
+
+                if (!string.IsNullOrEmpty(name))
+                {
+                    claims.Add(new Claim(SecurityConstants.AuthLevelKeyNameClaimType, name));
+                }
+
+                var identity = new ClaimsIdentity(claims, AuthLevelAuthenticationDefaults.AuthenticationScheme);
+                return AuthenticateResult.Success(new AuthenticationTicket(new ClaimsPrincipal(identity), Scheme.Name));
+            }
+            else
+            {
+                return AuthenticateResult.NoResult();
+            }
+        }
+
+        internal static async Task<(string, AuthorizationLevel)> GetAuthorizationKeyInfoAsync(HttpRequest request, ISecretManager secretManager)
```

Comment:
Where did all the tests for this stuff go? When this logic used to live in AuthorizationLevelAttribute, we had a battery of unit tests targeting that type. I can't find the tests anymore.